File: OPF_julia/create_ac.jl
using CSV
using DataFrames

"""
    create_ac(case_name::String) -> Dict{String,Any}

Load AC grid data from CSV files with the given `case_name` prefix.

Expected CSV files (located in the same directory):
  - `<case_name>_baseMVA_ac.csv`
  - `<case_name>_bus_ac.csv`
  - `<case_name>_branch_ac.csv`
  - `<case_name>_gen_ac.csv`
  - `<case_name>_gencost_ac.csv`
  - `<case_name>_res_ac.csv`

Returns a dictionary `ac` with the following keys:
  - "baseMVA"   : Scalar base MVA value.
  - "bus"       : Matrix containing AC bus data.
  - "branch"    : Matrix containing AC branch data.
  - "generator" : Matrix containing AC generator data.
  - "gencost"   : Matrix containing AC generator cost data.
  - "res"       : Matrix containing AC RES data.
"""
function create_ac(case_name::String)
    # Set the base path to the directory containing this file
    base_path = @__DIR__

    # Initialize the dictionary to hold AC network data
    ac = Dict{String,Any}()

    # List of required CSV file names
    required_files = [
        "$(case_name)_baseMVA_ac.csv",
        "$(case_name)_bus_ac.csv",
        "$(case_name)_branch_ac.csv",
        "$(case_name)_gen_ac.csv",
        "$(case_name)_gencost_ac.csv",
        "$(case_name)_res_ac.csv"
    ]

    # Check missing files
    missing_files = filter(f -> !isfile(joinpath(base_path, f)), required_files)
    if !isempty(missing_files)
        error("Missing required AC files: " * join(missing_files, ", "))
    end

    # Load each CSV file
    read_csv_matrix(file) = Matrix(DataFrame(CSV.File(joinpath(base_path, file); header=false)))
    baseMVA = read_csv_matrix("$(case_name)_baseMVA_ac.csv")
    ac["baseMVA"] = baseMVA[1, 1]
    ac["bus"]      = read_csv_matrix("$(case_name)_bus_ac.csv")
    ac["branch"]   = read_csv_matrix("$(case_name)_branch_ac.csv")
    ac["generator"] = read_csv_matrix("$(case_name)_gen_ac.csv")
    ac["gencost"]  = read_csv_matrix("$(case_name)_gencost_ac.csv")
    ac["res"] = read_csv_matrix("$(case_name)_res_ac.csv")

    return ac
end

File: OPF_cpp/create_ac.cpp
#include "create_ac.h"
#include "csv_reader.h"
#include <Eigen/Dense>
#include <unordered_map>
#include <string>
#include <iostream>
#include<fstream>
#include <stdexcept>

/*
 * Function: create_ac
 * -------------------
 * Loads AC grid data from CSV files based on the provided case name prefix.

 * Expected CSV files (located in the same directory):
 *   - <case_name>_baseMVA_ac.csv
 *   - <case_name>_bus_ac.csv
 *   - <case_name>_branch_ac.csv
 *   - <case_name>_gen_ac.csv
 *   - <case_name>_gencost_ac.csv
 *   - <case_name>_res_ac.csv

 * Return unordered_map with the following keys:
 *   "baseMVA"   : Scalar base MVA value.
 *   "bus"       : Matrix containing AC bus data.
 *   "branch"    : Matrix containing AC branch data.
 *   "generator" : Matrix containing AC generator data.
 *   "gencost"   : Matrix containing AC generator cost data.
 *   "res"       : Matrix containing RES data.
 */

std::unordered_map<std::string, Eigen::MatrixXd> create_ac(const std::string& case_name) {
    std::unordered_map<std::string, Eigen::MatrixXd> ac;

    // List of required CSV file names
    const std::string baseMVA_file = case_name + "_baseMVA_ac.csv";
    const std::string bus_file = case_name + "_bus_ac.csv";
    const std::string branch_file = case_name + "_branch_ac.csv";
    const std::string gen_file = case_name + "_gen_ac.csv";
    const std::string gencost_file = case_name + "_gencost_ac.csv";
    const std::string res_file = case_name + "_res_ac.csv";

    // Check missing files, if not, then save as the matrix
    try {
       
        std::ifstream ifs_baseMVA(baseMVA_file);
        if (!ifs_baseMVA.good()) {
            throw std::runtime_error("Missing required file: " + baseMVA_file);
        }
        ac["baseMVA"] = readCSVtoCpp(baseMVA_file);

        std::ifstream ifs_bus(bus_file);
        if (!ifs_bus.good()) {
            throw std::runtime_error("Missing required file: " + bus_file);
        }
        ac["bus"] = readCSVtoCpp(bus_file);

        std::ifstream ifs_branch(branch_file);
        if (!ifs_branch.good()) {
            throw std::runtime_error("Missing required file: " + branch_file);
        }
        ac["branch"] = readCSVtoCpp(branch_file);

        std::ifstream ifs_gen(gen_file);
        if (!ifs_gen.good()) {
            throw std::runtime_error("Missing required file: " + gen_file);
        }
        ac["generator"] = readCSVtoCpp(gen_file);

        std::ifstream ifs_gencost(gencost_file);
        if (!ifs_gencost.good()) {
            throw std::runtime_error("Missing required file: " + gencost_file);
        }
        ac["gencost"] = readCSVtoCpp(gencost_file);

        std::ifstream ifs_res(res_file);
        if (!ifs_res.good()) {
            throw std::runtime_error("Missing required file: " + res_file);
        }
        ac["res"] = readCSVtoCpp(res_file);
    }
    catch (const std::exception& e) {
        std::cerr << "Error in create_ac: " << e.what() << std::endl;
        throw std::runtime_error("Failed to load AC grid data. Please verify that the required CSV files exist.");
    }

    return ac;
}